### PR TITLE
Fix mps scaled dot attention

### LINKED ISSUE #149261: MPS Error: NDArray > 2^32 bytes in scaled_dot_product_attention
### 🐛 Describe the bug

An error, `/AppleInternal/Library/BuildRoots/d187755d-b9a3-11ef-83e5-aabfac210453/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:850: failed assertion [MPSNDArray initWithDevice:descriptor:isTextureBacked:] Error: total bytes of NDArray > 2**32'`, appears to occur in `out = torch.nn.functional.scaled_dot_product_attention(q, k, v)` when processing large tensors. The following is the minimal, reproducible code:
```python
import torch
device = torch.device("mps")
q = torch.randn(1, 12, 29640, 128).to(device)
k = torch.randn(1, 12, 29640, 128).to(device)
v = torch.randn(1, 12, 29640, 128).to(device)
out = torch.nn.functional.scaled_dot_product_attention(q, k, v)
print(out.shape)
```

### Versions

PyTorch version: 2.8.0.dev20250315
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.3.2 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.10.16 (main, Dec  3 2024, 17:27:57) [Clang 16.0.0 (clang-1600.0.26.4)] (64-bit runtime)
Python platform: macOS-15.3.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4 Max

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] torch==2.8.0.dev20250315
[pip3] torchaudio==2.6.0.dev20250315
[pip3] torchvision==0.22.0.dev20250315
[conda] Could not collect

### PR DESCRIPTION
**Pull request by rakshekaraj**
Fixes #149261

Previously, computations on large sequence lengths (seq_len > 12288) were failing due to excessive memory usage.

My solution:
Implemented Chunking for Memory-Efficient Computation:
-Instead of processing the full tensor in one step, we split operations into chunks (chunk_size = 4096).This avoids memory spikes.
-No compromise on FP32 accuracy, unlike previous FP16-based workarounds.

Updated scaled_dot_product_attention Logic
-Loop-based computation replaces single-step matrix multiplication.
-The output tensor is built incrementally to stay within memory limits.

Also tested and verified. Through chunking, I was successfully able to run seq_len=16384 without memory errors. Although chunking introduces a small computational overhead compared to running the full operation in one step, it prevents memory allocation failures

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @mcarilli @ptrblck @leslie-fang-intel

### CODE REVIEW

**Comment by HFDLYS** [comment]
I feel there might be some issues with the code, and I'm not sure if it's something on my end. I've tested the proposed fix extensively and observed a few problems.
## 1. **New Error Occurs for Certain `seq_len` Ranges in my code**  
While testing my code
```python
import torch
device = torch.device("mps")
def test_size1(seq_len):
	q = torch.randn(1, 12, seq_len, 128).to(device)
	k = torch.randn(1, 12, seq_len, 128).to(device)
	v = torch.randn(1, 12, seq_len, 128).to(device)
	out = torch.nn.functional.scaled_dot_product_attention(q, k, v)
	print(out.shape)

def test_size2(seq_len):
	q = torch.randn(1, 12, seq_len, 128).to(device)
	k = torch.randn(1, 12, 512, 128).to(device)
	v = torch.randn(1, 12, 512, 128).to(device)
	out = torch.nn.functional.scaled_dot_product_attention(q, k, v)
	print(out.shape)

test_size1(18720)
test_size2(18720)
```
I ran into a new error when `seq_len` falls within certain ranges. For instance, when `seq_len` is greater than 13500 but less than 26770, the computation fails with the following error message:

```
/AppleInternal/Library/BuildRoots/d187755d-b9a3-11ef-83e5-aabfac210453/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:829: failed assertion `[MPSNDArray initWithDevice:descriptor:isTextureBacked:] Error: NDArray dimension length > INT_MAX'.
```
## 2. **Functionality of `scaled_dot_product_attention` Possibly Not Fully Retained**  
While testing , I noticed an issue where the function fails under certain conditions. Specifically, if the `seq_len` of `q` differs from that of `k` or `v`, the function raises an error and does not complete the computation. This behavior deviates from the expected functionality of the original `scaled_dot_product_attention`, which should support such cases without issue, perhaps.

**Review by zou3519 — CHANGES_REQUESTED** [review]
probably not intentional?

**Comment by rakshekaraj** [comment]
Thank you for the review. I’ve looked into the issues you mentioned, and here’s what I’ve found:

1. Error for Certain seq_len Ranges
The error occurs when seq_len is between 13,500 and 26,770, triggering a Metal buffer size limit assertion.
This suggests that MPS is trying to allocate memory for an operation that exceeds INT_MAX in buffer size.
Solution: One option is to further adjust chunking strategy dynamically to avoid exceeding Metal’s internal buffer constraints.

2. Handling Mismatched seq_len Between q, k, and v
The current implementation assumes q, k, and v have the same seq_len, whereas PyTorch’s SDPA should allow k and v to have different lengths.
 Solution: Modify the chunking logic so that k and v are not chunked the same way as q. This would align better with the expected functionality.

I have tried to implement and make this work with my latest commit.

**Comment by rakshekaraj** [comment]
My fix helps us deal with varied seq_len but I am still working to prevent invalid buffer allocation errors for higher-length sequences.
